Explainability integration

NinaWie opened the pull request:
This is a working version for the backend for video prediction. 
Features:
- Overlay the heat maps
- Overlay a gauge showing the uncertainty (Moritz)
- Added support to upload either a video or an image
- Adapted app.py (tested)

Notes:
- we must take care of the app.py, because it is now modified in two branches --> The one here is the up to date version

jannisborn commented on pocovidnet/pocovidnet/app.py:
```diff
@@ -2,14 +2,31 @@
+import numpy as np
+
+# TODO: fill correct parameters for path to the weights
+# model = Evaluator(
```
This todo is obsolete no?

jannisborn commented on pocovidnet/pocovidnet/app.py:
```diff
@@ -19,12 +36,16 @@ def allowed_file(filename):
 
+def allowed_file_video(filename):
+    return '.' in filename and \
+           filename.split('.')[-1].lower() in ["mp4", "mpg", "mov", "mpeg", "avi"]
```
how about gifs? Are they impossible to parse? Also, this should be aligned with the frontend, i.e. the popup should only accept files that have a suff in this list. It may make sense to define this list somewhere globally to be sure that frontend and backend use the same formats.

NinaWie replied:
Good point, I added gifs

jannisborn commented on pocovidnet/pocovidnet/app.py:
```diff
@@ -19,12 +36,16 @@ def allowed_file(filename):
-
     In any case, 'source' and 'destination' should be given in URL.
     """
 
```
Are we sure that we always receive the filename, i.e. option 1 never applies? If yes, delete code, if no, we should detect whetther we are in O1 or O2.

jannisborn commented on pocovidnet/pocovidnet/app.py:
```diff
@@ -51,6 +72,69 @@ def predict():
+@app.route("/video_predict", methods=['POST'])
+def video_predict():
+    """
+    Debugging:
```
Short docstring here would be nice

jannisborn replied:
It's not clear how this function interacts with VideoEvaluator, especially the options to do/not do heatmap & confidence computation

NinaWie replied:
Yes the problem is that we don't have any checkboxes or selections in the frontend so far. Therefore the VideoEvaluator options are hard coded in the app.py so far, computing cams and epistemic uncertainty. We should add more options in the future in accordance with the frontend.

jannisborn commented on pocovidnet/pocovidnet/cam.py:
```diff
@@ -37,21 +37,32 @@ def get_class_activation_map(
-    if img.shape[1:3] != size:
-        raise ValueError(f'Img has size {img.shape}, should have {size}.')
+    #if img.shape[1:3] != size:
+    #    raise ValueError(f'Img has size {img.shape}, should have {size}.')
```
The default size is 224x224, what happens in case of other sizes like 1000x1000? We don't catch the error anymore where `img` has a size differing from `size`. Are we sure that the resizing below works in both directions, i.e. if img < size, but also size < img?

NinaWie replied:
Resizing works in both directions

jannisborn commented on pocovidnet/pocovidnet/evaluate_video.py:
```diff
@@ -73,64 +72,133 @@ def cam_important_frames(
+                otherwise one of 'epistemic' or 'aleatoric
         """
+        if uncertainty_method is not None:
+            cam_dims = (1000, 1000)
```
i'm confused on what the default is. 224 or 1000 pixels square?

NinaWie replied:
I changed the parameter name to cam_output_size, it has to be 1000 to display the gauge. The input image size (usually 224x224) is another parameter and it can be anything

jannisborn commented on pocovidnet/pocovidnet/evaluate_video.py:
```diff
@@ -73,64 +72,133 @@ def cam_important_frames(
+                self.models[model_idx],
+                class_idx,
+                return_map=False,
+                layer_name="block5_conv3",
```
ideally, `layer_name` should be configurable

NinaWie replied:
Done

jannisborn commented on pocovidnet/pocovidnet/utils_uncertainty.py:
```diff
@@ -0,0 +1,76 @@
+        os.path.dirname(os.path.realpath(__file__)), '../precision_overlays'
+    )
+
+    img_path = os.path.join(
```
I don't see those images in this PR. We could put them in a separate folder called assets (I think that's better than directly into plots)

NinaWie commented:
I deleted the notebooks, added the gauges (glad you saw that, my bad), and did some minor changes (see comments above). The build is failing though in pocovidscreen

jannisborn commented:
@NinaWie the Node.js build fails because NPM is down, see this issue from today: https://stackoverflow.com/questions/63280779/npm-err-404-not-found-get-https-registry-npmjs-org-typescript-eslint2fesl

I suggest to go ahead and merge since you didn't change the frontend.
